File: backend/app/models/quest.py
# backend/app/models/quest.py

"""Quest-related SQLAlchemy models."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import Enum

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    String,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.base_class import Base


class QuestCadence(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    LIMITED = "limited"


class QuestMetric(str, Enum):
    WORKOUTS_COMPLETED = "workouts_completed"
    ACTIVE_MINUTES = "active_minutes"


class QuestStatus(str, Enum):
    ACTIVE = "active"
    COMPLETED = "completed"
    CLAIMED = "claimed"
    EXPIRED = "expired"


class QuestTemplate(Base):
    __tablename__ = "quest_templates"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    code = Column(String(64), nullable=False, unique=True, index=True)
    title = Column(String(120), nullable=False)
    description = Column(String(255), nullable=True)
    cadence = Column(String(16), nullable=False, default=QuestCadence.DAILY.value)
    metric = Column(String(32), nullable=False)
    target_value = Column(Integer, nullable=False, default=1)
    reward_xp = Column(Integer, nullable=False)
    auto_claim = Column(Boolean, nullable=False, default=True)
    is_active = Column(Boolean, nullable=False, default=True)
    available_from = Column(
        DateTime(timezone=True),
        nullable=True,
        default=lambda: datetime.now(timezone.utc),
    )
    expires_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    instances = relationship(
        "UserQuest",
        back_populates="template",
        cascade="all, delete-orphan",
    )

    __table_args__ = (
        CheckConstraint(
            "cadence IN ('daily', 'weekly', 'limited')",
            name="ck_quest_templates_cadence",
        ),
        CheckConstraint(
            "metric IN ('workouts_completed', 'active_minutes')",
            name="ck_quest_templates_metric",
        ),
        CheckConstraint(
            "target_value > 0",
            name="ck_quest_templates_target_positive",
        ),
        CheckConstraint(
            "reward_xp > 0",
            name="ck_quest_templates_reward_positive",
        ),
    )


class UserQuest(Base):
    __tablename__ = "user_quests"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    template_id = Column(
        UUID(as_uuid=True),
        ForeignKey("quest_templates.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    status = Column(String(16), nullable=False, default=QuestStatus.ACTIVE.value)
    progress_value = Column(Integer, nullable=False, default=0)
    required_value = Column(Integer, nullable=False)
    cycle_start = Column(DateTime(timezone=True), nullable=False)
    cycle_end = Column(DateTime(timezone=True), nullable=True)
    available_from = Column(DateTime(timezone=True), nullable=False)
    expires_at = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    reward_claimed_at = Column(DateTime(timezone=True), nullable=True)
    last_progress_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    template = relationship("QuestTemplate", back_populates="instances")
    user = relationship("User", back_populates="quests")

    __table_args__ = (
        UniqueConstraint(
            "user_id",
            "template_id",
            "cycle_start",
            name="uq_user_quests_cycle",
        ),
        CheckConstraint(
            "status IN ('active', 'completed', 'claimed', 'expired')",
            name="ck_user_quests_status",
        ),
        CheckConstraint(
            "progress_value >= 0",
            name="ck_user_quests_progress_nonnegative",
        ),
        CheckConstraint(
            "required_value >= 0",
            name="ck_user_quests_required_nonnegative",
        ),
    )

File: frontend/lib/core/models/quest.dart
// frontend/lib/core/models/quest.dart

enum QuestCadence { daily, weekly, limited }

enum QuestMetric { workoutsCompleted, activeMinutes }

enum QuestStatus { active, completed, claimed, expired }

QuestCadence _parseCadence(String? value) {
  switch (value) {
    case 'weekly':
      return QuestCadence.weekly;
    case 'limited':
      return QuestCadence.limited;
    case 'daily':
    default:
      return QuestCadence.daily;
  }
}

QuestMetric _parseMetric(String? value) {
  switch (value) {
    case 'active_minutes':
      return QuestMetric.activeMinutes;
    case 'workouts_completed':
    default:
      return QuestMetric.workoutsCompleted;
  }
}

QuestStatus _parseStatus(String? value) {
  switch (value) {
    case 'completed':
      return QuestStatus.completed;
    case 'claimed':
      return QuestStatus.claimed;
    case 'expired':
      return QuestStatus.expired;
    case 'active':
    default:
      return QuestStatus.active;
  }
}

DateTime? _parseDate(dynamic value) {
  if (value == null) return null;
  if (value is String && value.isNotEmpty) {
    return DateTime.parse(value);
  }
  return null;
}

class QuestTemplateSummary {
  final String id;
  final String code;
  final String title;
  final String? description;
  final QuestCadence cadence;
  final QuestMetric metric;
  final int targetValue;
  final int rewardXp;
  final bool autoClaim;
  final DateTime? availableFrom;
  final DateTime? expiresAt;

  const QuestTemplateSummary({
    required this.id,
    required this.code,
    required this.title,
    required this.description,
    required this.cadence,
    required this.metric,
    required this.targetValue,
    required this.rewardXp,
    required this.autoClaim,
    required this.availableFrom,
    required this.expiresAt,
  });

  factory QuestTemplateSummary.fromJson(Map<String, dynamic> json) {
    return QuestTemplateSummary(
      id: json['id']?.toString() ?? '',
      code: json['code']?.toString() ?? '',
      title: json['title']?.toString() ?? '',
      description: json['description']?.toString(),
      cadence: _parseCadence(json['cadence']?.toString()),
      metric: _parseMetric(json['metric']?.toString()),
      targetValue: (json['target_value'] as num?)?.toInt() ?? 0,
      rewardXp: (json['reward_xp'] as num?)?.toInt() ?? 0,
      autoClaim: json['auto_claim'] == true,
      availableFrom: _parseDate(json['available_from']),
      expiresAt: _parseDate(json['expires_at']),
    );
  }
}

class QuestInstance {
  final String id;
  final QuestStatus status;
  final int progress;
  final int required;
  final double progressPct;
  final DateTime? availableFrom;
  final DateTime? expiresAt;
  final DateTime? cycleStart;
  final DateTime? cycleEnd;
  final DateTime? completedAt;
  final DateTime? rewardClaimedAt;
  final DateTime? lastProgressAt;
  final int rewardXp;
  final QuestTemplateSummary template;

  const QuestInstance({
    required this.id,
    required this.status,
    required this.progress,
    required this.required,
    required this.progressPct,
    required this.availableFrom,
    required this.expiresAt,
    required this.cycleStart,
    required this.cycleEnd,
    required this.completedAt,
    required this.rewardClaimedAt,
    required this.lastProgressAt,
    required this.rewardXp,
    required this.template,
  });

  bool get isCompleted =>
      status == QuestStatus.completed || status == QuestStatus.claimed;

  bool get isClaimed => status == QuestStatus.claimed;

  factory QuestInstance.fromJson(Map<String, dynamic> json) {
    final templateRaw = json['template'];
    if (templateRaw is! Map) {
      throw ArgumentError('Quest template is missing from quest payload');
    }
    return QuestInstance(
      id: json['id']?.toString() ?? '',
      status: _parseStatus(json['status']?.toString()),
      progress: (json['progress'] as num?)?.toInt() ?? 0,
      required: (json['required'] as num?)?.toInt() ?? 0,
      progressPct: (json['progress_pct'] as num?)?.toDouble() ?? 0.0,
      availableFrom: _parseDate(json['available_from']),
      expiresAt: _parseDate(json['expires_at']),
      cycleStart: _parseDate(json['cycle_start']),
      cycleEnd: _parseDate(json['cycle_end']),
      completedAt: _parseDate(json['completed_at']),
      rewardClaimedAt: _parseDate(json['reward_claimed_at']),
      lastProgressAt: _parseDate(json['last_progress_at']),
      rewardXp: (json['reward_xp'] as num?)?.toInt() ?? 0,
      template: QuestTemplateSummary.fromJson(
        Map<String, dynamic>.from(templateRaw),
      ),
    );
  }
}
